Breaking up sky rendering into parts

Abastro opened the pull request:
This PR breaks sky rendering into parts of sky layers, so that each part can be replaced individually.
Each sky layer can be registered/replaced/removed individually by name.

Also, each sky layer has their own parent - the parent forms a group of sky layers, making it easy to group similar layers.

Order can be specified between any two elements, so that one element comes before another.

There are convenience methods like registerBefore/registerAfter to conveniently add a layer before/after a layer.

Abastro committed:
A try to break up sky rendering into parts, which should be separate.

Gegy commented on patches/minecraft/net/minecraft/world/WorldProvider.java.patch:
```diff
@@ -65,12 +65,13 @@
  
 +    /*======================================= Forge Start =========================================*/
 +    private net.minecraftforge.client.IRenderHandler skyRenderer = null;
++    private net.minecraftforge.client.SkyRenderer theSkyRenderer = new net.minecraftforge.client.SkyRenderer();
```
The similarity of the names `skyRenderer` and `theSkyRenderer` could be misleading. Should they not be renamed to make the difference more clear?

williewillus replied:
yeah, also `the` as a prefix is an antiquated MCP-ism, and really shouldn't be there :P

Abastro replied:
Of course, I'll fix that.

Abastro committed:
Fixed field naming.

Abastro commented:
Why isn't this getting any tags?
Besides, I really want some input from modders changing sky rendering. Anyone?

Gegy commented on patches/minecraft/net/minecraft/world/WorldProvider.java.patch:
```diff
@@ -65,12 +65,13 @@
      }
  
 +    /*======================================= Forge Start =========================================*/
 +    private net.minecraftforge.client.IRenderHandler skyRenderer = null;
```
Could this be renamed to something like `legacySkyRenderer`? It's not public, so binary compatibility wouldn't be lost.

Abastro replied:
I see.

Gegy commented on patches/minecraft/net/minecraft/world/WorldProvider.java.patch:
```diff
@@ -118,19 +119,34 @@
 +        return 1.0;
 +    }
 +
++    @Deprecated
```
Should add a note here why it is deprecated, and how the new system can be used instead.

Abastro replied:
Okay.

Gegy commented on patches/minecraft/net/minecraft/world/WorldProvider.java.patch:
```diff
@@ -118,19 +119,34 @@
++
++    /** Sets sky renderer for certain pass and resource location. Renderer with higher priority is called earlier. */
++    @SideOnly(Side.CLIENT)
++    public void setSkyRenderer(net.minecraftforge.client.EnumSkyPass pass, net.minecraft.util.ResourceLocation id, @Nullable net.minecraftforge.client.IRenderHandler renderer, double priority)
```
Should it rather be forced that `setSkyRenderer` should be called on the `SkyRenderer`? The method of this name implies that it's setting a field, and things will be overridden if you do it twice. (even with a different pass)
Or maybe a different name would be suitable? Maybe `setSkyPassRenderer` or `setSkyRendererPass`?

Edit: As you can have more than one pass renderer, `add` might be better than `set`

Abastro replied:
I see, moving to SkyRenderer would definitely be better.

Gegy commented on src/main/java/net/minecraftforge/client/EnumSkyPass.java:
```diff
@@ -0,0 +1,36 @@
+
+package net.minecraftforge.client;
+
+public enum EnumSkyPass
```
From what I've seen, Forge has mostly been moving away from enums prefixed with `Enum`. Could be changed to `SkyRenderPass`

Abastro replied:
I didn't know that, thanks!

Gegy commented on src/main/java/net/minecraftforge/client/SkyRenderer.java:
```diff
@@ -0,0 +1,490 @@
+        RenderHandler(IRenderHandler handler, double priority)
+        {
+            this.handler = handler;
+            this.priority = priority;
```
Could priority be simplified to an enum of priority types, similar to how event priorities are handled?

Abastro replied:
No. It's priority like that applied on IWorldGenerator. It can't be replaced with enums because modders will want to add their own ordered renderers with arbitrary size.

Gegy replied:
👍

Gegy commented on src/main/java/net/minecraftforge/client/SkyRenderer.java:
```diff
@@ -0,0 +1,490 @@
+        @Override
+        public void render(float partialTicks, WorldClient world, Minecraft mc)
+        {
+            if (mc.world.provider.getDimensionType().getId() == 1)
```
These copied renderers could be hard to port if functionality changes in the original renderer. Would it be possible to call the vanilla renderer still, but filter what is rendered depending on the current pass?

Abastro replied:
Yes, it's possible. But I'll postpone the change till it's approved by modders doing sky rendering work. I guess this entire PR is moot before that.

Abastro commented:
I need actual input from modder who's working with sky rendering.
Is this PR acceptable? Or is there a reason you can't use this way of modifying the sky rendering?

Abastro committed:
Applied advices.

HellFirePvP commented:
> I need actual input from modder who's working with sky rendering.

I can't seem to find a way to remove existing bits of skyrender? Removing a specific skyrender based on RenderPass and key would be _desperately_ needed. Other than that, i can't see anything wrong with it from my current perspective. On that note, this is looking way better than skyrendering works atm... :P

covers1624 commented on patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch:
```diff
@@ -93,29 +93,23 @@
-+            return;
-+        }
-+
++        net.minecraftforge.fml.client.FMLClientHandler.instance().renderSky(p_174976_1_);
```
The return of this is never used, looks like the if got dropped when moving the logic.

Abastro replied:
Oops.

Abastro replied:
This is fixed now.

Abastro commented:
> I can't seem to find a way to remove existing bits of skyrender? Removing a specific skyrender based on RenderPass and key would be desperately needed.

I see, as renderers are already RenderPass-dependent I think I only have to add a method to unregister a sky renderer.

EDIT:
Now thinking about it, it's actully there; calling the register method with null renderer will remove the renderer. I think I need to special-case removing vanilla renderer, and it'll be enough.
Is it better with an unregister method?

Anyway, thanks for the feedback!

Abastro committed:
Fixed wording and added unregister vanilla method.

Gegy commented on src/main/java/net/minecraftforge/client/SkyRenderHandler.java:
```diff
@@ -53,19 +55,33 @@
+     * @param id the id to unregister
+     * @return the vanilla render handler, or <code>null</code> if there wasn't a vanilla renderer for specified pass-id pair
+     * */
+    public @Nullable IRenderHandler unregisterVanillaRenderer(SkyRenderPass pass, ResourceLocation id)
```
If there's only 1 vanilla renderer for each pass, does it make sense for the `id` to be passed in?

Abastro replied:
There are three vanilla renderers for SunMoonStars: the Sun, the Moon, and the Stars.

Gegy replied:
Ah, I see. Whoops.

mcenderdragon commented:
This looks good, it will defanetly make it easier to add custom objects to modded dimensions too.

Abastro committed:
Quick draft for vanilla part rendering.

Abastro committed:
Merge branch '1.12.x' into skyrendering

Abastro committed:
Added state calls for consistency when removed, and added the Test mod.

Abastro commented:
Added some GL state calls for consistency when the vanilla renderers are removed.
Added test mods which either removes either of sun/moon/stars, or gives red tint on either sun or moon.
Completed this PR, and I think I got enough feedback.

Now, how can I get this PR be approved and pulled?

Abastro commented:
Bump

LexManos commented:
Do not bump.

mezz commented:
If you are done requesting comments and ready to submit this as a feature PR, you can ping me and I will take a look. Otherwise I'll get around to it eventually as I look through all the PRs, like now.

At first look, please clean up the formatting. Patches should generally have braces on the same line to minimize patch lines, and non-patch code should always have braces on a new line.

Please update your original post to explain the PR as it is now, as a complete feature.

Abastro committed:
Fixed formatting.

Abastro commented:
Okay, @mezz. I fixed the formatting.

mezz commented on patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch:
```diff
@@ -93,29 +93,119 @@
-             this.func_180448_r();
-@@ -1470,6 +1492,7 @@
+-            this.func_180448_r();
++            if(renderPass == 0 || renderPass == 1) this.func_180448_r();
```
You should put `if (renderPass == 0 || renderPass == 1)` on a separate line above `this.func_180448_r();` to keep this patch simple.

mezz commented on patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch:
```diff
@@ -93,29 +93,119 @@
+         else if (this.field_72777_q.field_71441_e.field_73011_w.func_76569_d())
+         {
+-            GlStateManager.func_179090_x();
++            if(renderPass == 0 || renderPass == 1) GlStateManager.func_179090_x();
```
put the `if` as a line above `GlStateManager.func_179090_x();`

mezz commented on patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch:
```diff
@@ -93,29 +93,119 @@
-+            return;
-+        }
-+
++        int renderPass = net.minecraftforge.fml.client.FMLClientHandler.instance().renderSky(p_174976_1_);
```
Can you use `SkyRenderPass` instead of `int` here? I think it would make things clearer.

Abastro replied:
I can't, because there are several vanilla renderers for 'SunMoonStars' pass.

mezz commented on patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch:
```diff
@@ -93,29 +93,119 @@
+             }
+ 
+-            GlStateManager.func_179124_c(f, f1, f2);
++            if(renderPass == 0 || renderPass == 1) GlStateManager.func_179124_c(f, f1, f2);
```
`if` goes on newline before `GlStateManager.func_179124_c(f, f1, f2);`

mezz commented on patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch:
```diff
@@ -93,29 +93,119 @@
+                 GlStateManager.func_179121_F();
+                 GlStateManager.func_179103_j(7424);
+-            }
++            }}
```
put your brace on a new line, leave the existing line alone

mezz commented on patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch:
```diff
@@ -93,29 +93,119 @@
+             GlStateManager.func_179098_w();
+             GlStateManager.func_187428_a(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO);
+-            GlStateManager.func_179094_E();
++            GlStateManager.func_179094_E(); }
```
put your brace on a new line, leave the existing line alone

mezz commented on patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch:
```diff
@@ -93,29 +93,119 @@
+             GlStateManager.func_179131_c(1.0F, 1.0F, 1.0F, f16);
+             GlStateManager.func_179114_b(-90.0F, 0.0F, 1.0F, 0.0F);
+-            GlStateManager.func_179114_b(this.field_72769_h.func_72826_c(p_174976_1_) * 360.0F, 1.0F, 0.0F, 0.0F);
++            GlStateManager.func_179114_b(this.field_72769_h.func_72826_c(p_174976_1_) * 360.0F, 1.0F, 0.0F, 0.0F); }
```
put your brace on a new line, leave the existing line alone

mezz commented on patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch:
```diff
@@ -93,29 +93,119 @@
+             bufferbuilder.func_181662_b((double)f17, 100.0D, (double)f17).func_187315_a(1.0D, 1.0D).func_181675_d();
+             bufferbuilder.func_181662_b((double)(-f17), 100.0D, (double)f17).func_187315_a(0.0D, 1.0D).func_181675_d();
+             tessellator.func_78381_a();
++            GlStateManager.func_179121_F(); }
```
put your brace on a new line, leave the existing line alone

mezz commented on patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch:
```diff
@@ -93,29 +93,119 @@
+             bufferbuilder.func_181662_b((double)f17, -100.0D, (double)(-f17)).func_187315_a((double)f22, (double)f23).func_181675_d();
+             bufferbuilder.func_181662_b((double)(-f17), -100.0D, (double)(-f17)).func_187315_a((double)f24, (double)f23).func_181675_d();
+             tessellator.func_78381_a();
++            GlStateManager.func_179121_F(); }
```
put your brace on a new line, leave the existing line alone

mezz commented on patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch:
```diff
@@ -93,29 +93,119 @@
+                     GlStateManager.func_179148_o(this.field_72772_v);
+                 }
+             }
++            GlStateManager.func_179121_F(); }
```
put your brace on a new line, leave the existing line alone

mezz commented on patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch:
```diff
@@ -93,29 +93,119 @@
+             GlStateManager.func_179121_F();
+             GlStateManager.func_179098_w();
+-            GlStateManager.func_179132_a(true);
++            GlStateManager.func_179132_a(true); }
```
put your brace on a new line, leave the existing line alone

mezz commented on src/main/java/net/minecraftforge/client/SkyRenderHandler.java:
```diff
@@ -0,0 +1,217 @@
+        public int compareTo(RenderHandler o)
+        {
+            // Higher priority comes first
+            return -Comparator.<RenderHandler>comparingDouble(k -> k.priority).compare(this, o);
```
`Comparator` has a `reversed()` method for this purpose

mezz commented on src/main/java/net/minecraftforge/client/SkyRenderPass.java:
```diff
@@ -0,0 +1,36 @@
+    /**
+     * Sun, moon, stars and others.
+     * */
+    SUN_MOON_STARS,
```
Why are these combined into one pass?
Is it possible to draw over the stars and behind the moon (like a planet would)?
Is it possible to make this into a more general sorted system, so that mods can draw before or after any specific part?

Abastro replied:
I tried to keep what vanilla does. And I only got the three pass as they are necessary in most cases. While the Sun, the Moon and Stars are optional and even don't exist.
And yes, it's pretty tricky to draw over the stars and behind the moon in this system. I think I can solve this by making the priority be changed independently.
By sorted system, you mean something like mod sorting system? I think there is high probability where a cycle could happen. Also it's much more complex, so I don't think it's worth it.
EDIT: @mezz?

Abastro committed:
Fixed more formatting and fixed comparison issues.

Abastro commented:
@mezz, would you please check the replies?

Abastro commented:
So this is how a PR is ignored? Is this basically an unacceptable PR?

LexManos commented:
I typically leave my comments to private chats with Mezz as People think i'm to harsh. But hes on vacation and you insist on bumping your shit multiple times.
So speaking candidly, yes this PR is dumb, you split things but only partially and don't really split it to all the parts that it should be split into. Why are all the celestial objects in a single pass? is there any sorting system involved? 
Also some things are a bit confusing. You copy quite a bit of GLState code without actually telling what it does or where it comes from which causes issues with updating. You also use quite a bit of magic numbers, which is rather annoying.
Basically its a mess of confusing hack patches and considering I personally HATE rendering code i'm not inclined to clean it up for you. So i'm leaving it to Mezz, when he gets back.

Abastro commented:
So what mezz said is what you want to say, right?

I think it could be better when it's just left unassigned - anyone could have said 'This PR is messy pile of junk, your poor reasoning is like that as well' and I'd have tried to rewrite this one without a question. (I know some people to help me out, so it should be much better)

Above all, please tell me if you think this concept itself is wrong. Maybe this is the real reason? No point in putting effort in this case.

mezz commented:
Please update this PR to resolve conflicts, and run `gradle updateLicenses` to fix any new license headers.

Abastro committed:
Merge branch '1.12.x' of https://github.com/MinecraftForge/MinecraftForge into skyrendering

# Conflicts:
#	patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch
#	patches/minecraft/net/minecraft/world/WorldProvider.java.patch
#	src/main/java/net/minecraftforge/fml/client/FMLClientHandler.java

Abastro committed:
Rough outline of SkyLayer system

Abastro commented:
Resolved merge conflicts and draw a rough outline of SkyLayer system. How do you think?

MajorTuvok commented:
You said that you want a proper sorting system - the topological sort currently used for sorting mods maybe?

Abastro commented:
Yes, topological sort is what I'm going for.

Abastro committed:
Now with patches

Abastro committed:
Removed lines causing error just to make things build =P

Abastro committed:
Sketch of the layer graph system

Abastro committed:
Merge branch '1.12.x' of https://github.com/MinecraftForge/MinecraftForge into skyrendering

# Conflicts:
#	src/main/java/net/minecraftforge/fml/client/FMLClientHandler.java

Abastro committed:
Implementing topology sort for ordering

mezz commented:
Let me know when you are done updating this and I can take another look.

Abastro committed:
My brain stopped working here, what do?

Abastro committed:
Finished sorting mechanism

Abastro committed:
Nearly finished, only need to construct the test

Abastro commented:
@mezz I'm nearly finished updating it - only needs to work on the test code and performance now.
Would you review this PR once again? Now it revolves around tree of SkyLayer.
+ Discussed in discord, add registerBefore/registerAfter for convenience methods

Abastro committed:
registerBefore and registerAfter

Abastro committed:
Finished SkyRenderHandler

Abastro commented:
@mezz ? Please review

mcenderdragon commented:
Just a quick performance question: are the stars worth cashing or is the performance vs memory effect too small ?

Abastro commented:
@mcenderdragon Do you mean caching to VBO/display list? As far as I know, yes it is worth it - even though the effect is small.
I'll post a quick performance test when I finish the test.

mcenderdragon commented:
yes the VBO/display list.

liach commented:
Can you use the builtin topological sort library for the sky render layer list instead? Or is there any limitation with that library?

Abastro commented:
@liach what's the name of the library?

MajorTuvok commented:
It is not a library, it is part of forge (or fml to be precise).

The sole problem with it is, that the logs are hardcoded for sorting the modlist... 
Fixing that should not be difficult...

Abastro commented:
@MajorTuvok The usecase is quite different, so I can't use that.

MajorTuvok commented:
Why? The use cases are pretty similiar - you have sky layers which may depend on each other...

The sole effort is building the graph up correctly...

Abastro commented:
But my one has hierarchy, and should preserve previous order if no order is specified for consistency with vanilla.

Abastro committed:
Finished Test - now this works.

Abastro commented:
@mezz It's really finished now, along with the test. The only thing left is performance test.
Please review this time.

MajorTuvok commented:
Topological sorting is meant for cases where there are things dependant on each other - which is the case in a hierachy. 
And the default order will remain if specify your boundaries...

Abastro commented:
Yeah, I mean the sky layer could have several sublayers. And I found the topological sort in the part too connected with the mod sorting stuffs. I can't use it.
And how do you retain the default order?

MajorTuvok commented:
You can use it - f. e. adapting it to registries is possible and quite easy (I did that and then noticed that skyboy was already doing that and I then only applied the log fix to him).

You retain the default order by specifying it as boundaries. 
Give me 2 or 3 hours and I'll PR it to you - if I have the time to finish it, though I'll try.

liach commented:
You can add a secondary sort to a topological sort when parallel entries exist. In this case, the secondary sort is the default order.

MajorTuvok commented:
I just noticed that this is way more complex than I thought it is

liach commented:
Can you move the content of the update runnable into a private method and use method reference instead?

Abastro commented:
Yeah but I don't think the FML topology sort could make use of the secondary sort method easily.
And I'll switch to the method reference!

liach commented:
Why can't you use a regular Layer to Layer map in favor to a graph for the layer group dependencies? One layer has at most one parent group (as I see)

Abastro commented:
Because graph is more useful - it has many convenience methods

MajorTuvok commented:
ok, so I have implemented the toposort now.

I'll test it now (and to do that I have to get IntelliJ to apply your existing patches! Argh!).

MajorTuvok commented:
Something is really weird here - I can't start forge and trying to trick it by recompiling the Minecraft-codebase only works partially... somehow the SPacketEntityHeadLook and some others don't seem to exist. 
I think I did something wrong with the patches - I'll submit you a PR anyway.

MajorTuvok commented:
If you want to, you can take a look at it @Abastro, even though I uploaded it, without being able to test it.

liach commented:
IMO, a proper toposort for guava graph can be added. Probably one like this? 

https://github.com/KyoriPowered/lunar/blob/b5054bebfd98cb1f652645cf46dd51cb0771722d/src/main/java/net/kyori/lunar/graph/MoreGraphs.java#L72

MajorTuvok commented:
That would be quite helpful, so that one doesn't always have to use the fml toposort which might very well be removed/replaced in 1.13, because of the rewrite. Though I'm not informed about that.

Besides I like this iterative (hope that's the correct translation) approach...
XD

You'd then only have to throw in your implementation @Abastro...?

Abastro commented:
Closed an useless PR.

MajorTuvok commented:
Why useless?
I do see quite a few uses...

Daomephsta commented:
Yeah, this was shaping up nicely. It looked like it'd be a very versatile sky rendering system when it was done.

Abastro commented:
Oops. many people care, actually.

Abastro commented:
@liach How am I supposed to implement that? Do I need to copy the classes?

Abastro commented:
Also fixed the description for the current state - I don't care about markdown format, btw

Abastro commented:
@mezz I'm waiting for your feedbacks.

MajorTuvok commented:
@Abastro, liach send you a link to the toposort...
Just send your Graph through it... (in your build method...)

liach commented:
You can copy that code (it is mit licensed)

liach commented:
moreover, i'd ask @kashike if he allows forge to take the whole sort code and license it under lgpl (without referring to his lunar library)

liach commented:
rip.

Abastro commented:
Oh no, how did you notice..
I wanted to close this silently =/

Phylogeny commented:
@Abastro All 402 people watching this repo can get notifications when PRs/issues are opened/closed or when comments are added to them.

liach commented:
the close was on forge discord github channel, too

Abastro commented:
Sorry for bothering you again.
Anyway I lost the intent to continue this PR.
